Refactor Kconfig

careyk007 opened the pull request:
menuconfig should look a little nicer now. Screenshots below show the various menu screens. I went with a buried menu approach, but could change it to show/hide options on the main window if that would be better.

In addition to visual layout changes, I removed `ZMK_ACTION_MOD_TAP` since it was no longer used. I also added some code to enable the same config options used for BLE HID over GATT when just enabling BLE for Split communications. (i.e. `select BT`, `select BT_SMP`, etc. are selected for both `ZMK_BLE` and `ZMK_SPLIT_BLE`). I'm not sure if this was strictly necessary, but it doesn't seem to have broken anything.

![zmk_overview](https://user-images.githubusercontent.com/8930980/95701618-ea927e80-0bfe-11eb-9961-b87ad68cb0fe.png)
![zmk_menu](https://user-images.githubusercontent.com/8930980/95701625-efefc900-0bfe-11eb-94f2-e6c126fce07e.png)
![zmk_basic_setup](https://user-images.githubusercontent.com/8930980/95701634-f4b47d00-0bfe-11eb-818f-5f740396ba7a.png)
![zmk_split_support](https://user-images.githubusercontent.com/8930980/95701654-fd0cb800-0bfe-11eb-9f3b-9c0884cb8126.png)
![zmk_display_options](https://user-images.githubusercontent.com/8930980/95701659-026a0280-0bff-11eb-8d77-57cf74ba0052.png)
![zmk_power_management](https://user-images.githubusercontent.com/8930980/95701661-0564f300-0bff-11eb-8f64-6ac1751e0076.png)
![zmk_kscan_settings](https://user-images.githubusercontent.com/8930980/95701669-09911080-0bff-11eb-95b2-fb497f149700.png)


Should close #204.

careyk007 commented:
Ah shoot, checks are failing. Consider this WIP until further notice.

Mariappan commented:
That's so cool man :+1:

innovaker commented:
Thanks @careyk007! 👍   Looks good!  Maybe switch it from PR to Draft until it's ready for review?

careyk007 commented:
Checks are passing now, taking this out of draft.

petejohanson commented on app/Kconfig:
```diff
@@ -118,9 +94,17 @@ if ZMK_SPLIT
 
 config ZMK_SPLIT_BLE
 	bool "Split keyboard support via BLE transport"
-	depends on ZMK_BLE
```
I don't really agree with this tweak. Can you explain why this was necessary?

careyk007 replied:
I think this came from wanting a split keyboard that only interfaced with a host via USB. Each half would need to enable BLE for connecting to the other half, but not for connecting to a host. I may have misinterpreted what I needed to do to set that up though.

petejohanson replied:
We can explore that, but I'd rather not make that also part of the refactor here. My general philosophy is:

* Merge a pure refactor, with no additions/features.
* Build your work on top of the refactor as a follow up PR.

It allows keeping focus, and also helping identify the true source of regressions. It also helps reviewers understand scope of changes much more clearly.

petejohanson reviewed (commented):
In general, loving it! A few questions though on the specifics.

Thanks.

careyk007 commented on app/Kconfig:
```diff
@@ -9,34 +13,36 @@ config USB_DEVICE_PRODUCT
 	select USB_DEVICE_HID
 
-if USB
+if ZMK_USB
```
That makes sense. Switching to `ZMK_USB` allows it to be shown/hidden based on the selection of `USB` in menuconfig. I see the value of keeping it as `USB` though. Thinking through it, the Init Priorities (both USB and BLE) should probably be moved outside of basic keyboard setup since most users will not need to modify those defaults.

petejohanson replied:
Yeah, those init priorities really should be moved somewhere "advanced"

careyk007 commented:
@petejohanson Thanks for the feedback! I reverted my changes to `USB` and moved the initialization priorities for USB and BLE to the advanced page. 

I would like your thoughts on how to enable BLE for split communication while only allowing USB interaction with a host. I'm happy to revert my changes to `ZMK_SPLIT_BLE` if depending on `ZMK_BLE` is the best approach.

petejohanson reviewed (changes requested):
In general, this is great! I would ask that we revert the BLE/Split stuff for now, and keep this purely refactor focused, then worth through alternate USB/BLE topology support in a follow up.

Thanks!

careyk007 commented:
Thanks for the reviews! I have reverted my changes to `ZMK_SPLIT_BLE` as well as `build.yaml` to keep this a pure refactor.

petejohanson reviewed (approved):
Thanks!
